From 70dca412f03f9fbedddb89c9bbfd27752661a359 Mon Sep 17 00:00:00 2001
From: Arun Ravindran <arun.ravindran@intel.com>
Date: Mon, 19 Nov 2012 16:33:18 +0200
Subject: [PORT FROM MAIN] Telephony: Fix issue in automatic timezone update

BZ: 29186

Some networks provides the timezone offset including
daylight saving time but doesn't provide the daylight
saving time separately. Android framework compares
both the timezone offset and dst to fetch the correct
zone information. Since the NITZ provided DST doesn't
match with the one stored in device side, correct
zone information is not fetched resulting in automatic
timezone update not happening properly.

If the daylight saving time is not provided, then
force the android to fetch the zone by informing
that the daylight saving time is provided.

Change-Id: I9c49d30c251669600f70202dd1a79e29ae717146
Orig-Change-Id: Ic0e3c58038a42da5d1ac38e0cd05c73536ea686b
Category: aosp improvement
Domain: Telephony-Telephony-FWK
Origin: internal
Upstream-Candidate: yes
Signed-off-by: Jeevaka Badrappan <jeevaka.badrappan@intel.com>
Signed-off-by: Arun Ravindran <arun.ravindran@intel.com>
---
 .../telephony/gsm/GsmServiceStateTracker.java      |   13 ++++++++++++-
 1 files changed, 12 insertions(+), 1 deletions(-)

diff --git a/src/java/com/android/internal/telephony/gsm/GsmServiceStateTracker.java b/src/java/com/android/internal/telephony/gsm/GsmServiceStateTracker.java
index b13b442..ef42b24 100644
--- a/src/java/com/android/internal/telephony/gsm/GsmServiceStateTracker.java
+++ b/src/java/com/android/internal/telephony/gsm/GsmServiceStateTracker.java
@@ -1047,6 +1047,10 @@ final class GsmServiceStateTracker extends ServiceStateTracker {
                     } else {
                         zone = TimeUtils.getTimeZone(mZoneOffset, mZoneDst, mZoneTime, iso);
                         if (DBG) log("pollStateDone: using getTimeZone(off, dst, time, iso)");
+                        if (zone == null) {
+                            // Couldn't find a proper timezone.  Perhaps the DST data is wrong.
+                            zone = TimeUtils.getTimeZone(mZoneOffset,!mZoneDst, mZoneTime, iso);
+                        }
                     }
 
                     mNeedFixZoneAfterNitz = false;
@@ -1548,9 +1552,16 @@ final class GsmServiceStateTracker extends ServiceStateTracker {
 
                 if (mGotCountryCode) {
                     if (iso != null && iso.length() > 0) {
-                        zone = TimeUtils.getTimeZone(tzOffset, dst != 0,
+                        boolean isDstProvided = (dst != 0);
+                        zone = TimeUtils.getTimeZone(tzOffset, isDstProvided,
+                                c.getTimeInMillis(),
+                                iso);
+                        if (zone == null) {
+                            // Couldn't find a proper timezone.  Perhaps the DST data is wrong.
+                            zone = TimeUtils.getTimeZone(tzOffset, !isDstProvided,
                                 c.getTimeInMillis(),
                                 iso);
+                        }
                     } else {
                         // We don't have a valid iso country code.  This is
                         // most likely because we're on a test network that's
-- 
1.7.4.1

